Fix counter_cache decremented twice

malikolivier opened the pull request:
This commit fixes the bug reported in #31493 and #32079.

It includes the following test that would fail without the patch as
follows:

```
Failure:
BelongsToAssociationsTest#test_decrement_counter_cache_once [rails/activerecord/test/cases/associations/belongs_to_associations_test.rb:778]:
Expected: -1
  Actual: -2
```

The fix was proposed first in #31494 by "gm_zhou <gm_zhou@ctrip.com>".
I included a failing test case on top of his contribution.

malikolivier committed:
Fix counter_cache decremented twice

This commit fixes the bug reported in #31493 and #32079.

It includes the following test that would fail without the patch as
follows:

Failure:
BelongsToAssociationsTest#test_decrement_counter_cache_once
[rails/activerecord/test/cases/associations/belongs_to_associations_test.rb:778]:
Expected: -1
  Actual: -2

The fix was proposed in #31494 by "gm_zhou <gm_zhou@ctrip.com>".
The test was added by the author of this commit.

[Guangming Zhou, Malik Olivier Boussejra]

kaspth commented:
Hey, thanks for following up on this!

For good measure, can you use http://contributors.rubyonrails.org/faq#credit-several-people so the original author also gets credit?  :)

@kamipo can you take a look here?

malikolivier committed:
belongs_to_associations_test.rb: Simplify test case

Test case was using changing the a dog's nested attributes using
`accepts_nested_attributes_for`. However, this is not required to
reproduce the bug fixed by 6107b37.

Indeed, the bug can be reproduced simply by having a model with two different
`counter_cache`.

As we do not change the definitions of dogs or dog_lovers, the overall impact
of the PR over the project is reduced by this commit.

malikolivier commented:
@kaspth Thanks. I updated my commit message to giving credits according to the contribution guidelines you linked me too. Sorry I missed this step.

In the mean time, I updated the PR  to simplify the unit tests. The original PR was changing the definitions of two models, which was in fact not necessary to reproduce the bug. Hope it's better then way.

malikolivier commented:
@kamipo Any feedback? :) Will this fix still be released on the 5.2 branch?

kamipo commented:
I have been busy lately. I will visit again in weekend.

kaspth commented on activerecord/lib/active_record/associations/belongs_to_association.rb:
```diff
@@ -77,12 +77,20 @@ def require_counter_update?
 
+        def update_after_replace_counter_called
+          if (after_repace_counter_called = owner.instance_variable_get(:@_after_replace_counter_called))
+            after_repace_counter_called[reflection.foreign_key] = true
```
`repace` => `replace` here and above.

kaspth commented on activerecord/lib/active_record/associations/builder/belongs_to.rb:
```diff
@@ -34,8 +34,9 @@ def belongs_to_counter_cache_after_update(reflection)
-          if (@_after_replace_counter_called ||= false)
-            @_after_replace_counter_called = false
+          @_after_replace_counter_called ||= {}
+          if @_after_replace_counter_called.has_key?(foreign_key)
```
I thought `has_key?`was deprecated in favor of `key?`? I prefer the latter anyway.

kaspth commented on activerecord/test/cases/associations/belongs_to_associations_test.rb:
```diff
@@ -767,6 +769,18 @@ def test_replace_counter_cache
+    dog.update!(trainer: dog_lover_2)
+    dog_count_after = DogLover.find(previous_trainer_id).trained_dogs_count
+    assert_equal(-1, dog_count_after - dog_count_before)
+  end
```
I find this test unwieldy, can we simplify to this?

```ruby
def test_decrement_counter_cache_once
  original_trainer = DogLover.create!
  dog = Dog.create!(trainer: original_trainer)

  assert_difference -> { original_trainer.reload.trained_dogs_count }, -1 do
    dog.update!(trainer: DogLover.create!)
  end
end
```

malikolivier committed:
belongs_to_association.rb: Fix typo repace => replace

malikolivier committed:
belongs_to.rb: has_key? => key?

key? is preferable over has_key?

malikolivier committed:
belongs_to_associations_test.rb: Simplify test_decrement_counter_cache_once

malikolivier commented:
I rebased and fixed the conflicts.
